java/hiro-client/src/main/java/co/arago/hiro/client/rest/DefaultWebSocketClient.java

```diff
@@ -61,159 +190,120 @@ public DefaultWebSocketClient(String restApiUrl, String urlParameters, TokenProv
-                }
+    /**
+     * Reconnect a connection that should not have been closed. This also implements a delay strategy for repeated
+     * attempts to reconnect as suggested per RFC6455. Reconnect attempts only stop on {@link #close()} or an
+     * InterruptedException while sleeping.
+     */
+    private synchronized void reconnect() {
+        Duration nextTryDelay = Duration.ZERO;
 
-                process(loglistener, JSONValue.toJSONString(m));
+        closeWs();
 
-                if (WebsocketType.Event == type) {
-                    close();
-                    throw new HiroException("connection closed", 400);
+        while (true) {
+            try {
+                if (!running) {
+                    return;
                 }
 
-                if (running && !initial) {
-                    connect(false, initial);
-                }
-            }
+                long delay = nextTryDelay.getSeconds() * 1000;
+                Thread.sleep(delay);
 
-            @Override
-            public void onTextFrame(String payload, boolean finalFragment, int rsv) {
-                if (LOG.isLoggable(Level.FINEST)) {
-                    LOG.log(Level.FINEST, "received message " + payload);
+                if (LOG.isLoggable(Level.INFO)) {
+                    LOG.log(Level.INFO, "Reconnecting " + this);
                 }
 
-                if (handler != null) {
-                    handler.onTextFrame(payload, finalFragment, rsv);
-                }
+                connect(true);
 
-                Object o = JSONValue.parse(payload);
-                if (o instanceof Map && ((Map) o).containsKey("error")) {
-                    Map error = (Map) ((Map) o).get("error");
-                    if ((int) error.get("code") == 401) {
-                        tokenProvider.renewToken();
-                        connect(false, false);
-                        return;
-                    } else if (WebsocketType.Event == type) {
-                        throw new HiroException((String) error.get("message"), (int) error.get("code"));
-                    }
-                }
+                return;
+            } catch (HiroException ex) {
 
-                process(dataListener, payload);
-            }
+                LOG.log(Level.SEVERE, "Reconnect caught exception.", ex);
 
-            @Override
-            public void onPingFrame(byte[] payload) {
-                if (webSocketClient != null && webSocketClient.isOpen()) {
-                    webSocketClient.sendPongFrame(payload);
+                // Retry strategy
+                if (nextTryDelay.getSeconds() < 3) {
+                    nextTryDelay = nextTryDelay.plusSeconds(1);
+                } else if (nextTryDelay.getSeconds() < 60) {
+                    nextTryDelay = nextTryDelay.plusSeconds(new Random().nextInt(10) + 1);
+                } else {
+                    nextTryDelay = Duration.ofMinutes(new Random().nextInt(10) + 1);
```

Comment:
do you think it make sense to wait minutes if can not connect?

Reply:
What would you suggest? My strategy was to try to reconnect waiting 1,2,3 seconds, then a random number of seconds until the delay reaches one minute and then between one and ten minutes after that. The intent is to avoid cluttering the network with unnecessary requests when the socket has been gone for a long time.